File: week09/2/Prob1.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        # make neighbor list
        neighbors = dict()
        for n1, n2 in edges:
            if neighbors.get(n1, False):
                neighbors[n1].append(n2)
            else:
                neighbors[n1] = [n2]
            if neighbors.get(n2, False):
                neighbors[n2].append(n1)
            else:
                neighbors[n2] = [n1]
                


        
        # define BFS    
        q = deque([source])
        visited = set([source])
        while q:
            node = q.popleft()            
            if node == destination:
                return True            
            for n in neighbors[node]:
                if n not in visited:
                    visited.add(n)
                    q.append(n)
        # source에서 시작하는 loop에 destination이 없다면, path가 없음: False
        return False

File: week09/5/prob1.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {

        vector<vector<int>> adj = convertEdgesToAdj(n, edges);
        vector<bool> visited(n, false);

        return dfs_helper(source, destination, adj, visited);
    }

    vector<vector<int>> convertEdgesToAdj (int n, vector<vector<int>>& edges) {
        vector<vector<int>> adj(n);
        for (vector<int> edge : edges) {
            adj[edge.front()].push_back(edge.back());
            adj[edge.back()].push_back(edge.front());
        }
        return adj;
    }

    bool dfs_helper(int curr, int destination, const vector<vector<int>>& adj, vector<bool>& visited) {
        visited[curr] = true;

        // base case
        if (curr == destination) {
            return true;
        }
        
        // recursive case
        for (int neighbor : adj[curr]) {
            if (!visited[neighbor]) {
                if (dfs_helper(neighbor, destination, adj, visited)) {
                    return true;
                }
            }
        }
        return false;
    }
};
